### PR TITLE
externalresources(engine): refactor local file manager

### PR DESCRIPTION
**Pull request by CharlesCheung96**
### What problem does this PR solve?

Issue Number: ref #7114

### What is changed and how it works?
1. Add new interface [FileManager](https://github.com/pingcap/tiflow/pull/7127/files#diff-19c5de5a4ee5cc92546e9a176f86e9dfd080bd8b71936ce17e0a341c449b2408R54) and [ResourceDescriptor](https://github.com/pingcap/tiflow/pull/7127/files#diff-19c5de5a4ee5cc92546e9a176f86e9dfd080bd8b71936ce17e0a341c449b2408R54).
2. Refactor local file manager.
3. Use new interface in `broker` and `ResourceHandle` for easy integration with S3.
4. Add [S3Config](https://github.com/pingcap/tiflow/pull/7127/files#diff-501235c78fbd83c2364e208a6f767b7176753d6215b3c061b0fd71929e42ae6aR23)
5. Optimize layout.
```
.
├── broker
│   ├── broker.go
│   ├── broker_test.go
│   ├── interfaces.go
│   ├── local_broker.go
│   ├── main_test.go
│   ├── storage_handle.go
│   └── storage_handle_test.go
├── integration_test
│   ├── cluster_stub.go
│   ├── gc_test.go
│   └── mock_cluster.go
├── internal
│   ├── errors.go
│   ├── file_manager.go
│   ├── local
│   │   ├── file_manager.go
│   │   ├── file_manager_test.go
│   │   ├── file_utils.go
│   │   ├── main_test.go
│   │   ├── resource_controller.go
│   │   ├── resource_controller_test.go
│   │   └── resource_desc.go
│   ├── resource_controller.go
│   └── resource_desc.go
├── manager
│   ├── gc_coordinator.go
│   ├── gc_coordinator_test.go
│   ├── gc_runner.go
│   ├── gc_runner_test.go
│   ├── interfaces.go
│   ├── main_test.go
│   ├── mock_client.go
│   ├── service.go
│   ├── service_test.go
│   └── test_utils.go
└── model
    ├── config.go
    ├── main_test.go
    ├── model.go
    └── model_test.go

````
 
#### Tests 

 - Unit test
 - Integration test
 
### Release note 

```release-note
None`.
```

### CODE REVIEW

File: engine/pkg/externalresource/broker/broker.go [hunk 1]
```diff
@@ -117,12 +126,19 @@ func (b *DefaultBroker) RemoveResource(
 			fmt.Sprintf("unexpected resource type %s", tp))
 	}
 
+	fm := b.fileManagers[tp]
```
[the remark this answers is not in the document]
Comment (CharlesCheung96) [reply engine/pkg/externalresource/broker/broker.go:175]: Currently only localFileManager is supported, we will fix it later in another PR.

File: engine/pkg/externalresource/broker/broker.go [hunk 2]
```diff
@@ -217,3 +230,24 @@ func (b *DefaultBroker) checkForExistingResource(
...
+
+func (b *DefaultBroker) getPersistResource(
+	ctx context.Context, fm internal.FileManager,
+	record *resModel.ResourceMeta,
+	resName resModel.ResourceName,
+) (internal.ResourceDescriptor, error) {
+	ident := internal.ResourceIdent{
+		Name: resName,
+		ResourceScope: internal.ResourceScope{
+			ProjectInfo: tenant.NewProjectInfo("", record.ProjectID),
```
[the remark this answers is not in the document]
Comment (CharlesCheung96) [reply engine/pkg/externalresource/broker/broker.go:242]: Maybe we could add this field in another PR.

File: engine/pkg/externalresource/model/model.go [hunk 3]
```diff
@@ -157,3 +157,8 @@ func ParseResourcePath(rpath ResourceID) (ResourceType, ResourceName, error) {
 	suffix := path.Join(segments[1:]...)
 	return resourceType, suffix, nil
 }
+
+// ParseResourceID returns an ResourceID based on given ResourceType and ResourceName.
+func ParseResourceID(rtype ResourceType, name ResourceName) ResourceID {
```
Comment (amyangfei) [inline engine/pkg/externalresource/model/model.go:162]: The `Parse` word in this function name is not accurate. The function `ParseResourcePath` is more likely `PasreResourceID`.

File: engine/jobmaster/dm/worker_manager_test.go [hunk 4]
```diff
@@ -22,7 +22,7 @@ import (
 	"github.com/pingcap/log"
 	dmconfig "github.com/pingcap/tiflow/dm/config"
 	"github.com/pingcap/tiflow/engine/model"
-	resourcemeta "github.com/pingcap/tiflow/engine/pkg/externalresource/resourcemeta/model"
+
```
Comment (gozssky) [inline engine/jobmaster/dm/worker_manager_test.go:25]: Remove this blank line.

File: engine/pkg/orm/mock_test.go [hunk 5]
```diff
@@ -20,7 +20,8 @@ import (
 	"time"
 
 	frameModel "github.com/pingcap/tiflow/engine/framework/model"
-	resourcemeta "github.com/pingcap/tiflow/engine/pkg/externalresource/resourcemeta/model"
+
```
Comment (gozssky) [inline engine/pkg/orm/mock_test.go:23]: Remove this blank line.

File: engine/pkg/externalresource/model/model.go (hunk 3 above)
Comment (gozssky) [reply engine/pkg/externalresource/model/model.go:162]: How about `GenXXX` or `BuildXXX`?

File: engine/pkg/externalresource/internal/file_manager.go [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+	// Executor denotes the executor on which the resource is created.
+	Executor model.ExecutorID
+
+	// WorkerID denotes the worker that is creating or has created the resource.
+	// Note that it is NOT necessarily that consumes or depends on the resource.
+	WorkerID frameModel.WorkerID
+}
+
+// ResourceIdent provides information for the file manager to
+// uniquely determine where and how the resource is stored.
```
Comment (gozssky) [inline engine/pkg/externalresource/internal/file_manager.go:39]: The `WorkerID` should be unique in the system. Why do we need `ProjectInfo` and `ExecutorID` to identify a resource?

Comment (CharlesCheung96) [reply engine/pkg/externalresource/internal/file_manager.go:39]: This is a field reserved for multi-tenant resource isolation.

File: engine/pkg/externalresource/model/model.go [hunk 7]
```diff
@@ -133,8 +133,8 @@ const (
 	ResourceTypeS3        = ResourceType("s3")
 )
 
-// ParseResourcePath returns the ResourceType and the path suffix.
-func ParseResourcePath(rpath ResourceID) (ResourceType, ResourceName, error) {
+// GenResourcePath returns the ResourceType and the path suffix.
+func GenResourcePath(rpath ResourceID) (ResourceType, ResourceName, error) {
```
Comment (gozssky) [inline engine/pkg/externalresource/model/model.go:137]:
```suggestion
func PasreResourceID(rpath ResourceID) (ResourceType, ResourceName, error) {
```

**Review by gozssky — APPROVED** [review]

**Review by maxshuang — APPROVED** [review]

**Pull request merged**